### DISCUSSION TITLE
Force Plugin startup system to run before those from App

### DISCUSSION THREAD

**Discussion by AllanJRY** [bevyengine/bevy#8836] [Q&A] [answered]
Hello ! I have a question as I try to learn Bevy  !

A try to make a `Minesweeper`.
I started by making a **Plugin** with a startup system which generate a grid (it is in a bundle because I will try tu use this same bundle to make a snake game later).

Then after added the plugin to my App, I decide to add a startup system which drop some mines at random places on the generated grid.

But I'm façing a problem, the grid generation startup system and the mine dropping system are running in parallel,
so my dropping mines system has no cells because they aren't spawned yet !

I have to tell the startup system from my app to wait until my Plugin startup system has finished.

What a tried to do, is to define `SystemSet` to order my startup systems calls : 

I have a `GridSet` : 
```rust
#[derive(SystemSet, Debug, Hash, PartialEq, Eq, Clone)]
pub enum GridSet {
    Generation,
    GenerationFlush,
}
```

Add it to my **bundle** and configure it  : 
```rust
pub struct GridPlugin;

impl Plugin for GridPlugin {
    fn build(&self, app: &mut bevy::prelude::App) {
        app.insert_resource(GridParams::default())
            .configure_sets((
                GridSet::Generation,
                GridSet::GenerationFlush.after(GridSet::Generation),
            ))
            .add_startup_system(setup_grid.in_set(GridSet::Generation))
            .add_system(apply_system_buffers.in_set(GridSet::GenerationFlush));
    }
}

```

I made another set for the app, and configure it to run after my **GridSet::GenerationFlush**.

```rust
#[derive(SystemSet, Debug, Hash, PartialEq, Eq, Clone)]
enum SetupSet {
    CellStateInitialization,
}

fn main() {
    App::new()
        .add_plugin(GridPlugin)
        .configure_set(SetupSet::CellStateInitialization.after(GridSet::GenerationFlush))        
        .add_startup_system(setup_camera)
        .add_startup_system(drop_mines.in_set(SetupSet::CellStateInitialization))
        .run();
}
```

But `drop_mines` still be called in parallel, even if my grid generation system hasn't finished.

I think I'm doing something wrong, or there is something that I didn't understand clearly.. But after some search I couldn't find any answer which fix my issue, [This previous post](https://github.com/bevyengine/bevy/discussions/8335#discussion-5062287) seems to be exactly what I'm looking for, but even when I added `apply_system_buffers` system call in my `GridPlugin` , the issue still there.

Maybe I try to structure thing the wrong way..

**Comment by wilk10** [comment, marked as answer]
The following should work:
```rust
app.add_startup_systems((
    setup_grid,
    apply_system_buffers,
    drop_mines,
).chain()));
```
This way the three system are chained one after the other and are guaranteed to run in the specified order.

In your code above `apply_system_buffers` doesn't run during startup, but during the main schedule, so `drop_mines` runs before it during startup, and your grid won't be spawned yet.

**Reply by AllanJRY** [reply]
Oh yes ! This way it works thank you.

So, in the future, should I avoid having app startup systems depending on a bundle startup system ?

**Reply by wilk10** [reply]
Uhm i think you mean Plugin instead of Bundle? If I understand correctly.

Either way, no, it's fine, systems can be ordered against each other using sets or .before/.after even if they are added to the app via different plugins, even external ones.

**Reply by AllanJRY** [reply]
Aaah yes sorry.. you right I was talking about Plugin, I feel stupid. 

Thanks you for your help !

**Reply by wilk10** [reply]
It's absolutely normal! There's a lot of new terms to learn when starting off with Bevy.

Good luck with your project!
